File: 07/dotnet6/depths.cs
var path = System.Reflection.Assembly.GetExecutingAssembly().Location;
path = Path.GetDirectoryName(path);
var real = File.ReadAllText($"{path}\\Day07Input.txt");
var depths = Array.ConvertAll(real.Split(','), int.Parse).ToList();
Console.WriteLine($"Part 1 takes {Part1(depths)} fuel");
Console.WriteLine($"Part 2 takes {Part2(depths)} fuel");

int Part1(List<int> depths)
{
    // Work out the median for the provided depths
    int optimalDepth;
    depths.Sort();
    if (depths.Count % 1 == 1)
    {
        optimalDepth = (depths[(int)Math.Ceiling(depths.Count / 2.0)] + depths[(int)Math.Floor(depths.Count / 2.0)]) / 2;
    }
    else
    {
        optimalDepth = depths[depths.Count / 2];
    }
    // Return the sum of the differences from the optimal depth
    return depths.Sum(depth => Math.Abs(depth - optimalDepth));
}

int Part2(List<int> depths)
{
    List<int> fuelUsed = new();
    depths.Sort();
    for (var test = depths[0]; test <= depths[^1]; test++) // Loop through every possible depth
    {
        // Fuel used is a triangular number so add the sum of (x * x+1) / 2 to the list
        fuelUsed.Add(depths.Sum(depth => (Math.Abs(depth - test) * (1 + Math.Abs(depth - test))) / 2));
    }
    fuelUsed.Sort(); // Sort the fuel usage list
    return fuelUsed[0]; // and return the smallest
}

File: 07/python/depths.py
from math import ceil, floor
import os
import sys


def part1(depths):
    # Work out the median for the provided depths
    depths.sort()
    if (len(depths) % 1 == 1):
        optimal_depth = (depths[int(ceil(len(depths)/2))] +
                         depths[int(floor(len(depths)/2))])/2
    else:
        optimal_depth = depths[int(len(depths)/2)]
    # Return the sum of the differences from the optimal depth
    ret_val = 0
    for depth in depths:
        ret_val += abs(depth - optimal_depth)
    return ret_val


def part2(depths):
    fuel_used = []
    depths.sort()
    for test in range(depths[0], depths[-1]):
        fuel = 0
        for depth in depths:
            fuel += abs(depth - test) * (1 + abs(depth - test)) / 2
        fuel_used.append(fuel)
    fuel_used.sort()
    return int(fuel_used[0])


def main():
    file_path = os.path.join(os.path.dirname(__file__), "Day07Input.txt")
    with open(file_path, "+rt", encoding='utf-8-sig') as f:
        lines = f.readline()
    depths = list(map(int, lines.split(',')))
    print("Part 1 ", part1(depths))
    print("Part 2 ", part2(depths))


if __name__ == "__main__":
    sys.exit(int(main() or 0))
